File: src/serial_radio_transport_driver/ReliableMessenger.py
from datetime import datetime, timedelta
from serial_radio_transport_driver.SerialTransport import SerialRadioTransport
from serial_radio_transport_driver.Message import Message

class ReliableMessenger(object):

        def __init__(self, transport:SerialRadioTransport, retry_max:str = 4, ack_timeout_millis:int = 100):
            self._transport = transport
            self.set_retry(retry_max)
            self.set_ack_timeout(ack_timeout_millis)
            self._incomingMessage : Message = False
            self._newMessageFlag = False

        def available(self):
            #The flag is to ensure multiple calls to available() will not corrupt message
            if (self._newMessageFlag): return True

            #Receive the message and immediately send ACK.
            if (self._transport.available()):
                self._incomingMessage = self._transport.receive_message()
                self._newMessageFlag = True
                self._send_ack(self._incomingMessage)
                return True

            #Reachable if no message is available
            return False

        def receive_message(self) -> Message :
            if (self.available() == False): return None
            self._newMessageFlag = False
            return self._incomingMessage

        def send_message(self, message:Message) -> int:
            # Send the message
            self._transport.send_message(message)
            # Compute the expected ACK Checksum String
            expected_string = self._computeACKString(message)

            # Wait for returned ACK nessage and automatic retry
            # Try for max_retry number of times
            for retry_count in range(self._retry_max):
                # Within timeout peirod, try to read an incoming message
                time_out = datetime.utcnow() + timedelta(milliseconds=self._ack_timeout_millis)
                while (datetime.utcnow() < time_out):
                    if (self._transport.available()):
                        msg = self._transport.receive_message()
                        # If successful, return the number of retrys
                        if (msg.body == expected_string):
                            return retry_count
                        else:
                        # If unsuccessful, keep waiting for message
                            pass
                #Retry sending message again.
                if (retry_count < self._retry_max - 1 ):
                    self._transport.send_message(message)
            # When exhausted number of retrys, return -1
            return - self._retry_max

        def discard_unread_message(self, message:Message) -> None:
            _newMessageFlag = False
            while (self._transport.available()):
                self._transport.receiveMessage()

        def set_retry(self, retry_max:int) -> None:
            self._retry_max = retry_max

        def set_ack_timeout(self, millisec:int) -> None:
            self._ack_timeout_millis = millisec

        def _send_ack(self, original_message:Message) -> None:
            # Prepare the ack message
            ack_message = Message()
            ack_message.sender_address = self._transport.get_address()
            ack_message.receiver_address = original_message.sender_address
            ack_message.body = self._computeACKString(original_message)
            # Send message directly by transport
            self._transport.send_message(ack_message)

        def _computeACKString(self, original_message:Message) -> str:
            return "ACK" + chr(self.stringChecksum(original_message.body))


        def stringChecksum(self, original_string:str) -> int:
            c = 0
            index = 0
            for char in original_string:
                c ^= (ord(char) + index)    #Repeated Bitwise OR with (the char + index)
                index +=1
            return (c % 25) + 97            #Limits the returned byte to be ASCII a-z

if __name__ == "__main__":
    from serial import Serial
    import time

    #Prepare Serial Port
    serial_port = Serial('COM3', 115200, timeout=1)
    serial_port.reset_input_buffer()
    time.sleep(2)

    #Prepare SerialTransport
    transport = SerialRadioTransport(serial_port)
    transport.set_end_of_msg_char(chr(4))

    #Prepare Message 1 - Configure Radio Address
    msg = Message(chr(7),chr(7),'0a')
    transport.send_message(msg)

    #Prepare Message 2 - Configure Radio Freq
    msg = Message(chr(7),chr(7),'12')
    transport.send_message(msg)

    #Prepare ReliableMessenger
    messenger = ReliableMessenger(transport)

    #Send message
    messenger.set_retry(5)
    msg = Message('b','a','')
    resend_count = 0
    success_count = 0
    tries = 100
    for i in range(tries):
        #Send Message
        msg.body = "|abcd" + str(i) + "|"
        result = messenger.send_message(msg)
        #Counting if send is a success
        if (result >= 0):
            resend_count += result
            success_count +=1
        #print (i,end='\r')
        print ("Send message : "+ str (msg.body) +" retrys: " + str(result),end='\r')
        # #Receive Reply
        # correct_reply = False
        # time_out = datetime.utcnow() + timedelta(seconds=1)

        # while (datetime.utcnow() < time_out):
        #     if (correct_reply): break
        #     if (messenger.available()):
        #         reply = messenger.receive_message()
        #         print ("Reply: " + str(reply.body))
        #         if (reply.body == msg.body[::-1]):
        #             success_count +=1
        #             correct_reply = True
    print("Success: " + str(success_count) + " / "+ str(tries))
    print("Resend: " + str(resend_count))

File: src_cpp/libraries/reliable_messenger/src/reliable_messenger.cpp
/*
 Name:		reliable_messenger.cpp
 Created:	9/15/2019 12:49:36 PM
 Author:	leungp
 Editor:	http://www.visualmicro.com
*/


#if defined(ARDUINO) && ARDUINO >= 100
#include "arduino.h"
#else
#include "WProgram.h"
#endif

#include "reliable_messenger.h"

// cpp:class::Message
// -----------------------


int Message::length() const {
    return strlen(body);
}

// cpp:class::ReliableMessenger
// ------------------------------


ReliableMessenger::ReliableMessenger(Transport &transport) {
    _transport = &transport;
    _newMessageFlag = false;
}

// Checks if there is message waiting in the transport layer
// Automatically sends ACK when this is confirmed.
boolean ReliableMessenger::available() {
    // The flag is to ensure multiple calls to available() will not corrupt message
    if (_newMessageFlag) return true;
    // Receive the message and immediately send ACK.
    if (_transport->available()) {
        incomingMessage = * _transport->receiveMessage();
        _newMessageFlag = true;
        sendACK(& incomingMessage);
        return true;
    }
    //Reachable if no message is available
    return false;
}

const Message * ReliableMessenger::receiveMessage() {
    if (!available()) return NULL;
    _newMessageFlag = false;
    return & incomingMessage;
}

int ReliableMessenger::sendMessage(Message const * message) {
    //Send the message
    _transport->sendMessage(message);
    //Compute the expected ACK Checksum String
    char expectedString[5];
   computeACKString(message->body,expectedString);

    //Wait for ACK and automatic retry
    int retryCount;
    for (retryCount = 0; retryCount < _retryMax; retryCount++) {
        unsigned long startTime = millis();
        while (millis() - startTime < _sendSendTimeoutMillis) {
            // Try to read incoming message
            if (_transport->available()) {
                // Compare incoming message to expected ACK value
                Message msg = *_transport->receiveMessage();
                //Serial.print("IncomingMsg:");

                if (strcmp(msg.body, expectedString) == 0 ) {
                    //Serial.println("ACK Heard Send is Success");
                    return retryCount;
                } else {
                    //Serial.println("Bad Compare");

                }
            }
            // Confirm if the returned Message is correct
            // Return if the ACK message is correct
        }
        // Resend message
        if (retryCount < _retryMax - 1) _transport->sendMessage(message);
    }
    return -1;
}

// Discard unread messages from the Transport Layer
// returns: number of discard messages
unsigned int ReliableMessenger::discardUnreadMessages() {
    //Consume any unreceived messages from the transport layer.
    int count = 0;
    while (_transport->available()) {
        _transport->receiveMessage();
        count++;
    }
    //Consume any received but not read messages.
    _newMessageFlag = false; 
    return count;
}

// Sets the number of retry when sending message.
void ReliableMessenger::setRetry(int retryMax) {
    _retryMax = retryMax;
}

// Sets the duration of ACK timeout for sending message.
// This timeout applies to each retry.
void ReliableMessenger::setSendTimeout(unsigned long millisec) {
    _sendSendTimeoutMillis = millisec;
}

// Send ACK message back to the sender.
void ReliableMessenger::sendACK(Message * message) {
    Message ackMessage;
    ackMessage.senderAddress = _transport->getAddress();
    ackMessage.receiverAddress = message->senderAddress;
    //Construct the ACK message body
    char ackMessageBody[5];
    computeACKString(message->body, ackMessageBody);
    ackMessage.body = ackMessageBody;

    _transport->sendMessage(&ackMessage);
}

// Compute the ACK string according to a message content.
void ReliableMessenger::computeACKString(char * originalMessage, char * ackString) {
    strcpy(ackString, "ACK ");
    ackString[3] = stringChecksum(originalMessage);
}

// Compute checksum byte based on message content.
byte ReliableMessenger::stringChecksum(char *s) {
    byte c = 0;
    byte index = 0;
    while (*s != '\0') {
        c ^= ((*s++) + index); //Repeated Bitwise OR with (the char + index)
        index++;
    }
    return (c % 25) + 97; //Limits the returned byte to be ASCII a-z
}




// cpp:class::Transport 
// ----------------

// Sets the address 
void Transport::setAddress(byte address) {
    _address = address;
}

byte Transport::getAddress() {
    return _address;
}

// cpp:class::SerialTransport
// ----------------

SerialTransport::SerialTransport(Stream &stream, unsigned int bufferLength, byte address) {
    _receiveBuffer = new char[bufferLength]();
    _stream = &stream;
    setAddress(address); //Default address

}

SerialTransport::SerialTransport(Stream &stream, unsigned int bufferLength) {
    _receiveBuffer = new char[bufferLength]();
    _stream = &stream;

}

void SerialTransport::sendMessage(Message const * message) {

    _stream->write(message->body);
    _stream->write(_endOfMessageChar);
}

boolean SerialTransport::available() {
    //If Last Received Byte is a \0, then we should immediately return true without reading more data.
    if (_receiveIndex > 0) {
        if (_receiveBuffer[_receiveIndex - 1] == _endOfMessageChar) return true;
    }

    //Receive as much character as possible and return true if \0 is detected.
    if (_stream->available()) {
        char newChar = _stream->read();
        // End of message character - end char array with 0 and return true
        if (newChar == _endOfMessageChar) {
            _receiveBuffer[_receiveIndex] = 0;
            return true;
        }
        // Normal character - proceed to store in buffer
        else {
            _receiveBuffer[_receiveIndex] = newChar;
            _receiveIndex++;
        }
    }

    //Reachable when no \0 is detected and no characters are available from _stream.
    return false;
}

Message const * const SerialTransport::receiveMessage() {
    _receivedMessage.body = _receiveBuffer;
    _receivedMessage.receiverAddress = _address;
    _receivedMessage.senderAddress = 0U;
    _receiveIndex = 0; // reset Index
    return &_receivedMessage;
}

char SerialTransport::getEndOfMessageChar() {
    return _endOfMessageChar;
}

void SerialTransport::setEndOfMessageChar(char endOfMessageChar) {
    _endOfMessageChar = endOfMessageChar;
}


// cpp:class::SerialRadioTransport
// ----------------


SerialRadioTransport::SerialRadioTransport(Stream & stream, unsigned int bufferLength) :SerialTransport(stream, bufferLength) {
    //Calling only base class constructor
}

SerialRadioTransport::SerialRadioTransport(Stream & stream, unsigned int bufferLength, byte address) : SerialTransport(stream, bufferLength, address) {
    //Calling only base class constructor
}

void SerialRadioTransport::sendMessage(Message const * message) {
    _stream->write(message->receiverAddress);   //Byte 0
    _stream->write(message->senderAddress);     //Byte 1
    _stream->write(message->body);              //Byte 2 & onwards
    _stream->write(_endOfMessageChar);         //Byte final byte
}

boolean SerialRadioTransport::available() {
    //If Last Received Byte is a \0, then we should immediately return true without reading more data.
    if (_receiveIndex > 0) {
        if (_receiveBuffer[_receiveIndex - 1] == _endOfMessageChar) return true;
    }

    //Receive as much character as possible and return true if \0 is detected.
    while (_stream->available()) {
        char newChar = _stream->read();

        // If this is the first character, check if message is for me.
        if (_receiveIndex == 0) {
            if (newChar != _address) {
                messageForMe = false;   //Set messageForMe Flag . This stops subsquent storage.
            }
        }

        // End of message character - end char array with 0 and return true
        if (newChar == _endOfMessageChar) {
            if (messageForMe) {
                _receiveBuffer[_receiveIndex] = 0;  //Terminate char[]
                return true;
            } else { // Message is not for me but now it ends.
                _receiveIndex = 0;      //Reset Receive Index (This will consume the message)
                messageForMe = true;    //Reset messageForMe Flag
                continue;               //Skip the following storage in buffer
            }
        }
        // Normal character - proceed to store in buffer
        else {
            if (messageForMe) {
                _receiveBuffer[_receiveIndex] = newChar;
            }
        }
        _receiveIndex++;
    } // while (_stream->available()) 

    //Reachable when no \0 is detected and no characters are available from _stream.
    return false;
}

Message const * const SerialRadioTransport::receiveMessage() {
    _receivedMessage.receiverAddress = _receiveBuffer[0];
    _receivedMessage.senderAddress = _receiveBuffer[1];
    _receivedMessage.body = _receiveBuffer + 2;             // Message starts from byte 2
    _receiveIndex = 0;                                      // reset Index
    return &_receivedMessage;
}


// cpp:class::CC1101RadioTransport
// ----------------


//
//CC1101RadioTransport::CC1101RadioTransport(byte address) {
//    _address = address;
//
//}
//
//boolean CC1101RadioTransport::begin() {
//    return false;
//}
//
//void _CC1101RadioTransportMessageReceived() { CC1101RadioTransport::_messageReceived(); }
//
//void CC1101RadioTransport::sendMessage(Message const * message) {
//    //Detach Receive interrupt
//    detachInterrupt(_interruptPin);
//
//    //Construct new CCPACKET
//    CCPACKET packet;
//    packet.length = message->length();
//
//    //Copy message from serialRxBuffer to CCPACKET
//    strncpy((char *)packet.data, message->body, packet.length);
//
//    //Send message
//    _radio->sendData(packet);
//
//    //Print debug information to Serial
//
//    //Serial.print(F("(Sent packet,len="));
//    //Serial.print(serialRxLength);
//    //Serial.print(F(")"));
//    //Serial.write(SerialTernimation);
//
//    //Reattach Receive interrupt
//    attachInterrupt(_interruptPin, _CC1101RadioTransportMessageReceived, FALLING);
//
//}
//
//boolean CC1101RadioTransport::available() {
//    return boolean();
//}
//
//
